CORE-9041 - kill flows from flow pipeline by throwing `FlowMarkedForKillException`

conal-smith opened the pull request:
Allows flow killing to be driven by throwing a `FlowMarkedForKillException` from within the flow pipeline.

This triggers exception handling to convert the context to one that discards events and schedules cleanup for all open flow sessions.

"details" map captures any useful information about why a flow was terminated, in future this could be sent to counterparties indicating why their flow was killed.

A `KillFlowContextProcessor` creates a context for killing the flow, accompanied with unit test and injected into the flow pipeline.

Corda-api related change: https://github.com/corda/corda-api/pull/797, however, this is not a breaking change.

conal-smith committed:
CORE-9041 - add flow termination context to the flow pipeline

conal-smith committed:
CORE-9041 - add test for FlowGlobalPostProcessorImpl, and split termination context to its own file

conal-smith committed:
CORE-9041 - add test for flow event pipeline

conal-smith committed:
CORE-9041 - fix detekt

sprintengine commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/FlowEventContext.kt:
```diff
@@ -18,6 +18,7 @@ import net.corda.messaging.api.records.Record
  * @param outputRecords The [Record]s that should be sent back to the message bus when the pipeline completes.
  * @param T The type of [FlowEvent.payload].
  * @param mdcProperties properties to set the flow fibers MDC with.
+ * @param flowTerminatedContext The context of why this flow was termianted, if not null.
```
Just noticed a typo

sprintengine commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/impl/FlowGlobalPostProcessorImpl.kt:
```diff
@@ -143,4 +153,33 @@ class FlowGlobalPostProcessorImpl @Activate constructor(
+
+        return context.checkpoint.sessions
+            // not necessary to clean up sessions that are already closed or errored.
+            .filterNot { sessionState -> sessionState.status == SessionStateType.CLOSED || sessionState.status == SessionStateType.ERROR }
```
Do you remember @LWogan if we said we need to schedule cleanup for sessions that are "CLOSING"?

LWogan replied:
yes you need to schedule cleanup for CLOSING.

sprintengine commented on gradle.properties:
```diff
@@ -40,7 +40,7 @@ bouncycastleVersion=1.72
 # Change to 5.0.0.xx-SNAPSHOT to pick up maven local published copy
 #cordaApiVersion=5.0.0.xxx-SNAPSHOT
-cordaApiVersion=5.0.0.613-beta+
+cordaApiVersion=5.0.0.613-alpha-1673874971428
```
Will update this back to beta+ after https://github.com/corda/corda-api/pull/797/files is merged

sprintengine commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/FlowEventContext.kt:
```diff
@@ -26,7 +27,9 @@ data class FlowEventContext<T>(
+    val mdcProperties: Map<String, String>,
+    var flowTerminatedContext: FlowTerminatedContext? = null
+) {
+    fun isFlowToBeKilled() = flowTerminatedContext != null &&
```
I have a unit test for this.

conal-smith committed:
CORE-9041 - fix a few typos, add unit test for FlowEventContext

vkolomeyko commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/factory/impl/FlowMessageFactoryImpl.kt:
```diff
@@ -42,6 +42,13 @@ class FlowMessageFactoryImpl(private val currentTimeProvider: () -> Instant) : F
+    override fun createFlowKilledStatusMessage(checkpoint: FlowCheckpoint, details: Map<String, String>?): FlowStatus {
+        return getCommonFlowStatus(checkpoint).apply {
+            flowStatus =  FlowStates.KILLED
+            processingTerminationDetails =  details
```
Nit:
```suggestion
            flowStatus = FlowStates.KILLED
            processingTerminationDetails = details
```

vkolomeyko reviewed (commented):
LGTM
But needs input from @corda/flow-worker

sprintengine commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/impl/FlowEventPipelineImpl.kt:
```diff
@@ -51,6 +51,11 @@ class FlowEventPipelineImpl(
         }
 
```
So I know there are some integration tests (`*.AcceptanceTest`) and I think it might make more sense to add some of these tests when we add the validation into some of the handlers.

sprintengine commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/FlowEventContext.kt:
```diff
@@ -26,7 +27,9 @@ data class FlowEventContext<T>(
+    var flowTerminatedContext: FlowTerminatedContext? = null
+) {
+    fun isFlowToBeKilled() = flowTerminatedContext != null &&
+            flowTerminatedContext?.terminationStatus == FlowTerminatedContext.TerminationStatus.TO_BE_KILLED
```
We check that it's "TO_BE_KILLED" - there could be other termination statuses like `STRAY_EVENT` which don't need the flow to be killed, because it doesn't actually exist, the usage of this flag could look something like this draft PR of Simon's: https://github.com/corda/corda-runtime-os/pull/2796/files#diff-7db3307827af2a2e3f33ea48636e9f0272df864f549452e48fb384205b3ca33c

simon-johnson-r3 commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/impl/FlowEventPipelineImpl.kt:
```diff
@@ -51,6 +51,11 @@ class FlowEventPipelineImpl(
         }
 
     override fun eventPreProcessing(): FlowEventPipelineImpl {
+        if (context.isFlowToBeKilled()) {
```
It seems that in every case you want to exit the pipeline and only run the global post processor. Except I'm not sure you really even want to run that except to create a "kill" context. Did you instead try a pattern where you throw and catch and run an explict handler, e.g.

```
try {
  pipeline
    .eventPreProcessing()
    .runOrContinue(flowTimeout)
    .setCheckpointSuspendedOn()
    .setWaitingFor()
    .requestPostProcessing()
    .globalPostProcessing()
    .context
} catch (ex: MarkedForKillException) {
  pipeline.createKillFlowContext()
}
```
That would remove the need for multiple "if" statements at each stage.

I don't know if the code would end up better or not, just thought it might be worth a few mins refactoring to see what falls out.

Also here it feels like we're going through the motions of pipeline execution like a ritual rather for any benefit.

sprintengine replied:
Yeah you're right, there's nothing special about global post-processor that we can't do with a new function / processor on the pipeline to create flow kill contexts. When it comes to implementing validation on the operational statuses in some of the handlers (for example, validation on `p2pOperationalStatus` in `SendRequestHandler`, `ReceiveRequestHandler`) we can throw `MarkedForKillException` that'll get handled here, using exceptions to drive escaping out of the pipeline and killing the flow. I don't even think a flag on the context is required any more either.

conal-smith committed:
CORE-9041 - use an exception throwing model and processor on the pipeline to create "kill" contexts.

conal-smith committed:
CORE-9041 - use KeyValuePairList instead of map on avro object

conal-smith committed:
CORE-9041 - bump api

JamesHR3 commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/impl/FlowEventProcessorImpl.kt:
```diff
@@ -76,14 +77,18 @@ class FlowEventProcessorImpl(
-                    .requestPostProcessing()
-                    .globalPostProcessing()
-                    .context
+                try {
+                    pipeline
+                        .eventPreProcessing()
+                        .runOrContinue(flowTimeout)
+                        .setCheckpointSuspendedOn()
+                        .setWaitingFor()
+                        .requestPostProcessing()
+                        .globalPostProcessing()
+                        .context
+                } catch (e: FlowMarkedForKillException) {
+                    pipeline.createKillFlowContext(e.details)
+                }
```
Are we not better off folding this into the current error handling `try...catch` block rather than nesting them like this?

The other advantage is that we don't need to bake the kill flow context processor into the pipeline and can keep it separate. That probably removes some plumbing from the rest of this code.

sprintengine replied:
Yeah I'll put this logic alongside other exception processing:

```
} catch (e: FlowMarkedForKillException) {
    flowEventExceptionProcessor.process(e, pipeline.context)
}
```

LWogan replied:
You're not throwing `FlowMarkedForKillException` anywhere in this pr. Do you need to pass the details around via the exception? is it not available to the pipeline already. or could you just set it there. Hard to tell without seeing how you're planning on getting `details`

sprintengine replied:
Use case is some handler throws the exception with some details about why... e.g. SendAndReceive handler validates virtual node's p2pOperationalStatus is ACTIVE, otherwise `throw FlowMarkedForKillException("p2p operational status is inactive...")`

Though I'm starting to think all the details is not necessary, possibly just a simple "reason" String is all

JamesHR3 commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/impl/KillFlowContextProcessorImpl.kt:
```diff
@@ -0,0 +1,59 @@
+        )
+    }
+
+    private fun createFlowMapperSessionKilledEvents(context: FlowEventContext<Any>, now: Instant): List<Record<*, FlowMapperEvent>> {
```
This is triggering sessions to be cleaned up if they haven't already completed right? I think we should probably error any outstanding sessions at this point as well, as otherwise we're relying on the other side to notice that we've cleaned the session up early.

LWogan replied:
@lankydan Did this integration so he might be better to ask. But the pattern i think he was following was that you need to interact with something to find out if it is ERROR'd. However, there is one handler in the code for FlowIORequest.SubFlowFailed where it eagerly sends out a SessionError when subflow fails. Otherwise it will be the heartbeating mechanism that will fail when the Flow Mapper cleans up its state. 

We have a ticket open to finish off the mapper logic. As at the moment it will swallow all messages and not reply with anything when its in state CLOSING so currently that heartbeating would probably fail a bit quicker than waiting for the state to go to null.

LWogan replied:
https://r3-cev.atlassian.net/browse/CORE-8962

i think we should implement this ticket. This would allow us to signal to the other side that this session is ERROR'd after we clean up. If we wanted to signal that this flow failed due to being killed we could add some additional states like KILLED or expand the FlowMapperState to store a little context about why the session was errored rather than just an enum

additionally: I think we should eagerly send the error messages before nulling the checkpoint. and if the message gets lost in p2p land ,the mapper will eventually cause the other side to error anyway.

sprintengine replied:
Now we're sending errors to all flow sessions too

conal-smith committed:
CORE-9041 - move the exception handling to the exception processor for flow kill exceptions
- additionally set flow checkpoint to "deleted"
- additionally send out error messages to flow sessions that aren't CLOSED or ERRORed already.
- schedule cleanup events for any sessions that haven't already been scheduled, even if ERRORed.

conal-smith committed:
CORE-9041 - change details map to a string

conal-smith committed:
CORE-9041 - bump api

LWogan commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/impl/FlowEventExceptionProcessorImpl.kt:
```diff
@@ -154,6 +164,39 @@ class FlowEventExceptionProcessorImpl @Activate constructor(
+            val now = Instant.now()
+            val checkpoint = context.checkpoint
+
+            val activeSessionIds = checkpoint.sessions.filterNot { sessionState ->
```
i think we need to be a little more careful here. The FlowCheckpoint getters such as checkpoint.sessions will throw exceptions if theres no sessions associated with this checkpoimt. If this happens the withEscalation  closure will catch it and DLQ the record but the rest of your logic in this handler will be missed.

LWogan commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/impl/FlowEventProcessorImpl.kt:
```diff
@@ -76,14 +77,14 @@ class FlowEventProcessorImpl(
-                    .requestPostProcessing()
-                    .globalPostProcessing()
-                    .context
+                    pipeline
```
indentation :D

conal-smith committed:
CORE-8610 Add check on checkpoint existing first before acting on checkpoint's internal state (sessions)

conal-smith committed:
CORE-8610 fix post rebase

LWogan reviewed (approved):
LGTM
